File: src/main/java/com/nugurang/service/TaskService.java
package com.nugurang.service;

import com.nugurang.constant.ProgressName;
import com.nugurang.dao.PositionDao;
import com.nugurang.dao.ProgressDao;
import com.nugurang.dao.TaskDao;
import com.nugurang.dao.UserDao;
import com.nugurang.dao.WorkDao;
import com.nugurang.dao.XrefTaskPositionDao;
import com.nugurang.dao.XrefUserTaskDao;
import com.nugurang.dto.TaskInputDto;
import com.nugurang.entity.TaskEntity;
import com.nugurang.entity.XrefTaskPositionEntity;
import com.nugurang.entity.XrefUserTaskEntity;
import java.util.Optional;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class TaskService {
    private final PositionDao positionDao;
    private final ProgressDao progressDao;
    private final TaskDao taskDao;
    private final UserDao userDao;
    private final WorkDao workDao;
    private final XrefTaskPositionDao xrefTaskPositionDao;
    private final XrefUserTaskDao xrefUserTaskDao;

    @Transactional
    public TaskEntity createTask(TaskInputDto taskInputDto, Long work) {
        TaskEntity taskEntity = taskDao.save(TaskEntity.builder().name(taskInputDto.getName()).order(Optional.ofNullable(taskInputDto.getOrder()).orElseGet(() -> Optional.ofNullable(taskDao.findFirstByOrderByOrderDesc()).map(prevTaskEntity -> prevTaskEntity.getOrder() + 1).orElse(0))).difficulty(Optional.of(taskInputDto.getDifficulty()).orElse(1)).work(workDao.findById(work).get()).progress(Optional.ofNullable(taskInputDto.getProgress()).map(progressId -> progressDao.findById(progressId).get()).orElseGet(() -> Optional.ofNullable(progressDao.findByName(ProgressName.TODO.name())).get())).build());
        xrefUserTaskDao.saveAll(taskInputDto.getUsers().stream().map(userId -> userDao.findById(userId).get()).map(userEntity -> XrefUserTaskEntity.builder().user(userEntity).task(taskEntity).build()).collect(Collectors.toList()));
        xrefTaskPositionDao.saveAll(taskInputDto.getPositions().stream().map(positionId -> positionDao.findById(positionId).get()).map(positionEntity -> XrefTaskPositionEntity.builder().task(taskEntity).position(positionEntity).build()).collect(Collectors.toList()));
        return taskEntity;
    }

    public Optional<TaskEntity> getTask(Long taskId) {
        return taskDao.findById(taskId);
    }

    public TaskEntity updateTask(TaskInputDto taskInputDto, Long taskId) {
        TaskEntity taskEntity = taskDao.findById(taskId).get();
        taskEntity.setName(taskInputDto.getName());
        if (Optional.ofNullable(taskInputDto.getOrder()).isPresent()) taskEntity.setOrder(taskInputDto.getOrder());
        if (Optional.ofNullable(taskInputDto.getDifficulty()).isPresent()) taskEntity.setDifficulty(taskInputDto.getDifficulty());
        if (Optional.ofNullable(taskInputDto.getProgress()).isPresent()) {
            taskEntity.setProgress(progressDao.findById(taskInputDto.getProgress()).get());
        }
        return taskDao.save(taskEntity);
    }

    public Long deleteTask(Long taskId) {
        taskDao.deleteById(taskId);
        return taskId;
    }

    //<editor-fold defaultstate="collapsed" desc="delombok">
    @SuppressWarnings("all")
    
    public TaskService(final PositionDao positionDao, final ProgressDao progressDao, final TaskDao taskDao, final UserDao userDao, final WorkDao workDao, final XrefTaskPositionDao xrefTaskPositionDao, final XrefUserTaskDao xrefUserTaskDao) {
        this.positionDao = positionDao;
        this.progressDao = progressDao;
        this.taskDao = taskDao;
        this.userDao = userDao;
        this.workDao = workDao;
        this.xrefTaskPositionDao = xrefTaskPositionDao;
        this.xrefUserTaskDao = xrefUserTaskDao;
    }
    //</editor-fold>
}

File: src/main/kotlin/com/nugurang/service/TaskService.kt
package com.nugurang.service

import com.nugurang.annotation.DaoOp
import com.nugurang.constant.ProgressName
import com.nugurang.dao.*
import com.nugurang.dto.TaskInputDto
import com.nugurang.entity.*
import com.nugurang.exception.NotFoundException
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class TaskService(
    private val userService: UserService,
    private val positionDao: PositionDao,
    private val progressDao: ProgressDao,
    private val taskDao: TaskDao,
    private val workDao: WorkDao,
    private val xrefTaskPositionDao: XrefTaskPositionDao,
    private val xrefUserTaskDao: XrefUserTaskDao,
) {
    @Transactional
    fun createTask(taskInputDto: TaskInputDto, work: Long): TaskEntity {
        val taskEntity = taskDao.save(
            TaskEntity(
                name = taskInputDto.name,
                order = taskInputDto.order
                    ?: (taskDao.findFirstByOrderByOrderDesc()?.let { it.order + 1 } ?: 0),
                difficulty = taskInputDto.difficulty ?: 1,
                work = workDao.findByIdOrNull(work) ?: throw NotFoundException(WorkEntity::class.java),
                progress = taskInputDto.progress?.let {
                    progressDao.findByIdOrNull(it) ?: throw NotFoundException(ProgressEntity::class.java)
                }
                ?: (progressDao.findByName(ProgressName.TODO.name) ?: throw NotFoundException(ProgressEntity::class.java))
            )
        )

        xrefUserTaskDao.saveAll(
            taskInputDto.users.map { userService.getUser(it) }
            .map { XrefUserTaskEntity(user = it, task = taskEntity) }
        )

        xrefTaskPositionDao.saveAll(
            taskInputDto.positions.map {
                positionDao.findByIdOrNull(it) ?: throw NotFoundException(PositionEntity::class.java)
            }
            .map { XrefTaskPositionEntity(task = taskEntity, position = it) }
        )

        return taskEntity
    }

    fun getTask(taskId: Long): TaskEntity {
        return taskDao.findByIdOrNull(taskId) ?: throw NotFoundException(TaskEntity::class.java)
    }

    @DaoOp
    fun updateTask(taskInputDto: TaskInputDto, taskId: Long): TaskEntity {
        val taskEntity = taskDao.findById(taskId).get()
        taskEntity.name = taskInputDto.name
        taskInputDto.order?.let { taskEntity.order = it }
        taskInputDto.difficulty?.let { taskEntity.difficulty = it }
        taskInputDto.progress?.let {
            taskEntity.progress = progressDao.findByIdOrNull(it) ?: throw NotFoundException(ProgressEntity::class.java)
        }

        return taskDao.save(taskEntity)
    }

    @DaoOp
    fun deleteTask(taskId: Long): Long {
        taskDao.deleteById(taskId)
        return taskId
    }
}